clippy_lints/src/swap.rs

```diff
@@ -262,3 +264,82 @@ fn check_suspicious_swap(cx: &LateContext<'_>, block: &Block<'_>) {
         }
     }
 }
+
+/// Implementation of the xor case for `MANUAL_SWAP` lint.
+fn check_xor_swap(cx: &LateContext<'_>, block: &Block<'_>) {
+    for window in block.stmts.windows(3) {
+        if_chain! {
+            if let Some((lhs0, rhs0)) = extract_sides_of_xor_assign(&window[0]);
+            if let Some((lhs1, rhs1)) = extract_sides_of_xor_assign(&window[1]);
+            if let Some((lhs2, rhs2)) = extract_sides_of_xor_assign(&window[2]);
+            if eq_expr_value(cx, lhs0, rhs1);
+            if eq_expr_value(cx, rhs1, lhs2);
+            if eq_expr_value(cx, rhs0, lhs1);
+            if eq_expr_value(cx, lhs1, rhs2);
+            then {
+                let span = window[0].span.to(window[2].span);
+                let mut applicability = Applicability::MachineApplicable;
+                match check_for_slice(cx, lhs0, rhs0) {
+                    Slice::Swappable(slice, idx0, idx1) => {
+                        if let Some(slice) = Sugg::hir_opt(cx, slice) {
+                            span_lint_and_sugg(
+                                cx,
+                                MANUAL_SWAP,
+                                span,
+                                &format!(
+                                    "this xor-based swap of the elements of `{}` can be \
```

Comment:
I think we can safely simply the message like this, and do similarly with the other message. But you may disagree.

```suggestion
                                    "this xor-based swap can be \
```

Reply:
I unified messages. So now it is: "This looks like you are swapping elements of `foo` manually."

Reply:
Hmm I actually think that mentioning the method `slice::swap` is the most important part. The error is: "this method is available and you are not using it".

Reply:
It is being mentioned. The full message is:
```
error: this looks like you are swapping elements of `foo` manually
  --> $DIR/swap.rs:91:5
   |
LL | /     foo[0] ^= foo[1];
LL | |     foo[1] ^= foo[0];
LL | |     foo[0] ^= foo[1];
   | |_____________________^ help: try: `foo.swap(0, 1)`
```

In-line with:
```
error: this looks like you are swapping elements of `foo` manually
  --> $DIR/swap.rs:36:5
   |
LL | /     let temp = foo[0];
LL | |     foo[0] = foo[1];
LL | |     foo[1] = temp;
   | |_________________^ help: try: `foo.swap(0, 1)`
```
I'm not in favor of changing old code too much. But if message isn't clear we can change it.